File: docker/golang/03-http-fanout/main.go
// Go + Docker (runc) — HTTP fan-out HTTP service (I/O-bound experiment).
//
// On each inbound request, dispatch N concurrent outbound HTTP GETs to the
// in-cluster `io-echo` backend; the backend sleeps for delay_ms before
// responding, so the handler is dominated by outbound I/O wait rather than
// CPU work. Concurrent outbound is via sync.WaitGroup + goroutines over a
// shared http.Client with a tuned transport — the idiomatic Go I/O pattern.
//
// This is the docker-runc counterpart to the wasm-tinygo Spin component; they
// expose the same /fanout HTTP API to keep the comparison apples-to-apples.

package main

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"os"
	"strconv"
	"sync"
	"time"
)

const (
	backendURL     = "http://io-echo.http-fanout.svc.cluster.local/echo"
	defaultN       = 5
	maxN           = 20
	defaultDelayMs = 50
	maxDelayMs     = 1000
)

// ── Response ─────────────────────────────────────────────────────────────────
type FanoutResponse struct {
	Runtime   string `json:"runtime"`
	N         int    `json:"n"`
	DelayMs   int    `json:"delay_ms"`
	OkCount   int    `json:"ok_count"`
	ErrCount  int    `json:"err_count"`
	ElapsedUs int64  `json:"elapsed_us"`
	Responses []int  `json:"responses,omitempty"`
}

// Shared client — tuned transport for keep-alive against the io-echo backend.
var httpClient = &http.Client{
	Timeout: 5 * time.Second,
	Transport: &http.Transport{
		MaxIdleConns:        128,
		MaxIdleConnsPerHost: 64,
		IdleConnTimeout:     90 * time.Second,
	},
}

// ── Handler (I/O-bound fan-out) ──────────────────────────────────────────────
func fanoutHandler(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()

	n := defaultN
	if raw := q.Get("n"); raw != "" {
		if v, err := strconv.Atoi(raw); err == nil && v > 0 {
			if v > maxN {
				v = maxN
			}
			n = v
		}
	}
	delayMs := defaultDelayMs
	if raw := q.Get("delay_ms"); raw != "" {
		if v, err := strconv.Atoi(raw); err == nil && v >= 0 {
			if v > maxDelayMs {
				v = maxDelayMs
			}
			delayMs = v
		}
	}
	noList := q.Get("no_list") == "true" || q.Get("no_list") == "1"

	url := fmt.Sprintf("%s?delay_ms=%d&size_b=256", backendURL, delayMs)

	start := time.Now()

	statuses := make([]int, n)
	var wg sync.WaitGroup
	wg.Add(n)
	for i := 0; i < n; i++ {
		i := i
		go func() {
			defer wg.Done()
			req, err := http.NewRequestWithContext(r.Context(), http.MethodGet, url, nil)
			if err != nil {
				statuses[i] = 0
				return
			}
			resp, err := httpClient.Do(req)
			if err != nil || resp == nil {
				statuses[i] = 0
				return
			}
			defer resp.Body.Close()
			statuses[i] = resp.StatusCode
		}()
	}
	wg.Wait()

	okCount, errCount := 0, 0
	for _, s := range statuses {
		if s >= 200 && s < 300 {
			okCount++
		} else {
			errCount++
		}
	}

	elapsed := time.Since(start).Microseconds()

	slog.Info("fanout completed",
		"n", n,
		"delay_ms", delayMs,
		"elapsed_us", elapsed,
		"ok_count", okCount,
		"err_count", errCount,
	)

	resp := FanoutResponse{
		Runtime:   "golang-docker",
		N:         n,
		DelayMs:   delayMs,
		OkCount:   okCount,
		ErrCount:  errCount,
		ElapsedUs: elapsed,
	}
	if !noList {
		resp.Responses = statuses
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(resp); err != nil {
		slog.Error("encode error", "err", err)
	}
}

func healthHandler(w http.ResponseWriter, _ *http.Request) {
	w.WriteHeader(http.StatusOK)
}

// ── Entry point ──────────────────────────────────────────────────────────────
func main() {
	logger := slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
		Level: slog.LevelInfo,
	}))
	slog.SetDefault(logger)

	mux := http.NewServeMux()
	mux.HandleFunc("GET /fanout", fanoutHandler)
	mux.HandleFunc("GET /health", healthHandler)

	addr := ":8080"
	slog.Info("server starting", "addr", addr)

	if err := http.ListenAndServe(addr, mux); err != nil {
		slog.Error("server error", "err", err)
		os.Exit(1)
	}
}

File: docker/rust/03-http-fanout/src/main.rs
// Rust + Docker (runc) — HTTP fan-out HTTP service (I/O-bound experiment).
//
// On each inbound request, dispatch N concurrent outbound HTTP GETs to the
// in-cluster `io-echo` backend; the backend sleeps for delay_ms before
// responding, so the handler is dominated by outbound I/O wait rather than
// CPU work. Concurrent outbound is via `futures::future::join_all` over a
// shared `reqwest::Client` — the idiomatic Rust async I/O pattern.
//
// This is the docker-runc counterpart to the wasm-rust Spin component; they
// expose the same /fanout HTTP API to keep the comparison apples-to-apples.

use axum::{
    extract::{Query, State},
    http::StatusCode,
    response::Json,
    routing::get,
    Router,
};
use futures::future::join_all;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::Instant;
use tracing::info;

const BACKEND_URL: &str = "http://io-echo.http-fanout.svc.cluster.local/echo";

const DEFAULT_N: usize = 5;
const MAX_N: usize = 20;
const DEFAULT_DELAY_MS: u64 = 50;
const MAX_DELAY_MS: u64 = 1000;

// ── Query params ─────────────────────────────────────────────────────────────
#[derive(Deserialize)]
struct FanoutParams {
    n: Option<usize>,
    delay_ms: Option<u64>,
    #[serde(default)]
    no_list: Option<u8>,
}

// ── Response ─────────────────────────────────────────────────────────────────
#[derive(Serialize)]
struct FanoutResponse {
    runtime:    &'static str,
    n:          usize,
    delay_ms:   u64,
    ok_count:   usize,
    err_count:  usize,
    elapsed_us: u128,
    #[serde(skip_serializing_if = "Option::is_none")]
    responses:  Option<Vec<u16>>,
}

// ── Handler (I/O-bound fan-out) ──────────────────────────────────────────────
async fn fanout_handler(
    State(client): State<Arc<reqwest::Client>>,
    Query(params): Query<FanoutParams>,
) -> Json<FanoutResponse> {
    let n = params.n.unwrap_or(DEFAULT_N).clamp(1, MAX_N);
    let delay_ms = params.delay_ms.unwrap_or(DEFAULT_DELAY_MS).min(MAX_DELAY_MS);
    let no_list = params.no_list.unwrap_or(0) != 0;

    let url = format!("{BACKEND_URL}?delay_ms={delay_ms}&size_b=256");

    let start = Instant::now();

    let futures_iter = (0..n).map(|_| {
        let client = client.clone();
        let url = url.clone();
        async move { client.get(&url).send().await }
    });
    let results = join_all(futures_iter).await;

    let mut ok_count = 0usize;
    let mut err_count = 0usize;
    let mut statuses: Vec<u16> = Vec::with_capacity(n);
    for r in &results {
        match r {
            Ok(resp) => {
                let s = resp.status().as_u16();
                statuses.push(s);
                if (200..300).contains(&s) {
                    ok_count += 1;
                } else {
                    err_count += 1;
                }
            }
            Err(_) => {
                err_count += 1;
                statuses.push(0);
            }
        }
    }

    let elapsed_us = start.elapsed().as_micros();
    info!(n, delay_ms, elapsed_us, ok_count, err_count, "fanout completed");

    Json(FanoutResponse {
        runtime:    "rust-docker",
        n,
        delay_ms,
        ok_count,
        err_count,
        elapsed_us,
        responses:  if no_list { None } else { Some(statuses) },
    })
}

async fn health_handler() -> StatusCode {
    StatusCode::OK
}

// ── Entry point ──────────────────────────────────────────────────────────────
#[tokio::main]
async fn main() {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::from_default_env()
                .add_directive("http_fanout=info".parse().unwrap()),
        )
        .json()
        .init();

    let client = Arc::new(
        reqwest::Client::builder()
            .pool_max_idle_per_host(64)
            .build()
            .expect("reqwest client"),
    );

    let app = Router::new()
        .route("/fanout", get(fanout_handler))
        .route("/health", get(health_handler))
        .with_state(client);

    let addr = "0.0.0.0:8080";
    info!("listening on {addr}");
    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
